libopenage/screenshot.cpp

```diff
@@ -46,44 +52,89 @@ std::string ScreenshotManager::gen_next_filename() {
 
 
 void ScreenshotManager::save_screenshot() {
+	job::JobManager *job_mgr = this->job_manager;
+
+	uint32_t *pxdata = new uint32_t [this->window_size.x * this->window_size.y];
+	glReadPixels(0, 0,
+	             this->window_size.x, this->window_size.y,
+	             GL_RGBA, GL_UNSIGNED_BYTE, pxdata);
+
+	encode_png(pxdata);
+//	job_mgr->enqueue<void>(&encode_png);
+//asdf:	clean unneccessary includes/variables
+}
+
+void ScreenshotManager::encode_png(uint32_t* pxdata) {
+	uint32_t **rows = nullptr;
```

Comment:
If libpng only works with rows, the better choice here would be a `std::vector<uint32_t*> row_ptrs`. 

``` cpp
row_ptrs.reserve(this->window_size.y);
for (size_t i = 0; i < this->window_size.y; i++) {
    row.push_back(&pxdata[i * this->window_size.x]);
}
// as vector is guaranteed to be a continuous memory block:
png_set_rows(png_ptr, info_ptr, static_cast<png_bytepp>(&rows[0]));
```

I guess png can do this internally though, and you can just give it the pxdata buffer and dimensions.

Reply:
what's the difference between static_cast and ()-cast?
```
libopenage/screenshot.cpp:131:70: error: invalid static_cast from type ‘unsigned int**’ to type 'png_bytepp {aka unsigned char**}’
  png_set_rows(png_ptr, info_ptr, static_cast<png_bytepp>(&row_ptrs[0]));
                                                                      ^```

Reply:
See the `Explanation 1)`: http://en.cppreference.com/w/cpp/language/explicit_cast

The cast fails because the typecheck for `static_cast` detects something invalid here, as an int to char cast is platform dependent (byteorder).

Reply:
so, what's the fix? and since it's an **int to **char cast, the byteorder shouldn't be a problem?

Reply:
Hm, I'm not sure if the buffer shouldn't be a `char` array from the beginning. otherwise the int byte order may be flipped on platforms. Just reinterpreting it as something different doesn't solve the problem.

Reply:
I don't understand exactly what could be flipped. I understood from [this](http://en.cppreference.com/w/cpp/language/reinterpret_cast) under 'type aliasing' in the last bullet point that it's ok. not sure though, if this refers to our situation. I will change it though.

Reply:
If we first store the pixels in an `int == int32_t` with gl, it will use some byte order:
```
// GL_UNSIGNED_BYTE denotes that the input is unsigned bytes, so the type should be uint8_t[]
// if we pass int32 there, it will still work and written in order of RGBA
// if one of the ints was read, it would be 0xAABBGGRR
glReadPixels(0, 0, width, height, GL_RGBA, GL_UNSIGNED_BYTE, pxdata.get());
```

If we pass ints there and let libpng read the lines as `char`s, the pixels will again be read `rgba`, so my flipping thing was wrong. But the underlying data type is wrong, the memory is accessed only as `uint8_t`, so it should be a `uint8_t`.